src/robot/running/keywordrunner.py

```diff
@@ -206,16 +256,99 @@ def _to_number_with_arithmetics(self, item):
             raise TypeError("Expected number, got %s." % type_name(item))
         return number
 
-    def _run_one_round(self, data, values):
-        name = ', '.join(format_assign_message(var, item)
-                         for var, item in zip(data.variables, values))
-        result = KeywordResult(kwname=name,
-                               type=data.FOR_ITEM_TYPE)
-        for var, value in zip(data.variables, values):
-            self._context.variables[var] = value
-        runner = KeywordRunner(self._context, self._templated)
-        with StatusReporter(self._context, result):
-            runner.run_keywords(data.keywords)
+
+class ForInZipLoopRunner(BasicForLoopRunner):
+    def __init__(self, context, templated=False):
+        super(ForInZipLoopRunner, self).__init__(context, templated)
+
+    def _flavor_name(self):
+        return 'IN ZIP'
+
+    def _replace_variables(self, data):
+        """
+        Make sure we have as many loop-variables as zipped-iterables.
+
+        (Other than that, delegate to the superclass.)
+        """
+        values = super(ForInZipLoopRunner, self)._replace_variables(data)
+        if len(data.variables) == len(data.values):
+            return values
+        raise DataError("Expected %d Loop variables, but found %d"
+                % (len(data.values), len(data.variables)))
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
```

Comment:
- What does `_validate()` refer to in the comment? The main reason I don't like comments (or unnecessary docs) is that they get outdated so easily.
- Errors have used `FOR IN RANGE` style earlier. Should use that here too (or change those places to use `For-In-Range` style instead).
- Should use `utils.type_name(item)` instead of `item.__class__` for various reasons:
  - Get nicer name with known types.
  - Get consistent name with different Python interpreters (i.e. avoid problems with tests).
  - Avoid exception with types that don't actually have `__class__`.
- We try to keep line lengths below 80 chars. That's not so exact but this is too long.
- We haven't had a note like `but not strings` when talking about list-like objects elsewhere in the code base. No need to have it here either.
- As already noted, we generally have `.` at the end of the error.

Perhaps something simple like this would be good:

    "FOR IN ZIP items must all be list-like, got %s." % type_name(item)